[package] implement `get_resource_reader` API

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51676 [package] make importer a little more obscure
* **#51674 [package] implement `get_resource_reader` API**

See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

suo committed:
[package] implement `get_resource_reader` API

See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

[ghstack-poisoned]

suo committed:
Update on "[package] implement `get_resource_reader` API"


See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

[ghstack-poisoned]

suo commented on torch/package/importer.py:
```diff
@@ -448,3 +459,48 @@ def patched_getfile(object):
+
+    def open_resource(self, resource):
+        from io import BytesIO
+        return BytesIO(self.importer.load_binary(self.fullname, resource))
```
I think possibly we should make the `importer` API look more like the resource_reader API as well (so, `read_binary`, `open_binary` instead of `load_binary`). But for now, I just wrap it in a bytestream to conform with the interface

zdevito commented on torch/package/importer.py:
```diff
@@ -448,3 +459,48 @@ def patched_getfile(object):
+        path = self.importer._zipfile_path(self.fullname, name)
+        return self.importer.zip_reader.has_record(path)
+
+    def contents(self):
```
If this is written another way it can avoid having to add boolean flags to `_zipfile_path`, which I think it is an improvement:

```
filename = self.fullname.replace('.', '/')
for f in get_all_records()
   match = re.match(f"^{quote(self.filename)}/([^/]+)$", f)
  if match:
    yield match[0]
```

suo replied:
tried this, but in order get conformance with resource_reader API the regex gets complicated enough to not be immediately clear (in order to capture subdirs, future `/` slashes are optionally allowed but are not part of the match group). Seems more readable for future people to leave it as is imo, plus this is the implementation I stole from cpython lol

zdevito commented on torch/package/importer.py:
```diff
@@ -215,6 +215,14 @@ def get_source(self, module_name) -> str:
+            package = self._get_package(fullname)
+        except ImportError:
+            return None
+        assert package.__loader__ is self
```
In this case, should be return `None`. Otherwise It assert fail if some asks for resources for "sys", for instance.

suo committed:
Update on "[package] implement `get_resource_reader` API"


See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

[ghstack-poisoned]

suo committed:
Update on "[package] implement `get_resource_reader` API"


See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

[ghstack-poisoned]

suo committed:
Update on "[package] implement `get_resource_reader` API"


See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

[ghstack-poisoned]

suo committed:
Update on "[package] implement `get_resource_reader` API"


See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

[ghstack-poisoned]

suo committed:
Update on "[package] implement `get_resource_reader` API"


See
https://docs.python.org/3/library/importlib.html#importlib.abc.ResourceReader

Differential Revision: [D26237034](https://our.internmc.facebook.com/intern/diff/D26237034)

[ghstack-poisoned]

zdevito reviewed (approved):
looks good!
